datasets/imagenet/imagenet.py

```diff
@@ -0,0 +1,148 @@
+            description=_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "image": datasets.Image(),
+                }
+            ),
+            homepage=_HOMEPAGE,
+            citation=_CITATION,
+        )
+
+    def _prep_validation(self, data_dir, val_prep_script):
+        with open(val_prep_script, "r") as f:
+            commands = f.readlines()
+
+        for idx, command in enumerate(commands):
+            command = command.strip()
+            if "mkdir" in command:
+                folder = command.split(" ")[-1]
+                folder = os.path.join(data_dir, "val", folder)
+                os.makedirs(folder, exist_ok=True)
+            # mkdir will definitely happen before move in continuous incremental loop.
+            elif "mv" in command:
+                splits = command.split(" ")
+                image_file = splits[1]
+                folder = splits[2]
+                dst_path = os.path.join(data_dir, "val", folder)
+                if os.path.exists(os.path.join(dst_path, image_file)):
+                    continue
+                src_path = os.path.join(data_dir, "val", image_file)
+                if not os.path.exists(src_path):
+                    continue
+                shutil.move(src_path, dst_path)
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators."""
+        data_dir = os.path.abspath(os.path.expanduser(dl_manager.extract(dl_manager.manual_dir)))
+
+        if not os.path.exists(data_dir):
+            raise FileNotFoundError(
+                f"{data_dir} does not exist. Make sure you insert a manual dir via `datasets.load_dataset('matinf', data_dir=...)` that includes files unzipped from the MATINF zip. Manual download instructions: {self.manual_download_instructions}"
+            )
+        val_prep_script = dl_manager.download(_VAL_PREP_SCRIPT)
+        images_path = os.path.join(data_dir, "ILSVRC", "Data", "CLS-LOC")
+
+        self._prep_validation(images_path, val_prep_script)
+
+        label_classes = os.listdir(os.path.join(images_path, "train"))
+
+        self.info.features = datasets.Features(
+            {"image": datasets.Image(), "labels": datasets.ClassLabel(names=sorted(label_classes))}
+        )
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={"files": dl_manager.iter_files(os.path.join(images_path, "train"))},
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.VALIDATION,
+                gen_kwargs={"files": dl_manager.iter_files(os.path.join(images_path, "val"))},
+            ),
+            # datasets.SplitGenerator(
+            #     name=datasets.Split.TEST,
+            #     gen_kwargs={"files": dl_manager.iter_files(os.path.join(images_path, "test")), "no_labels": True},
+            # ),
+        ]
+
+    def _generate_examples(self, files, no_labels=False):
```

Comment:
`no_labels` is currently used for test set but I will remove it if we can't figure out a way to make `labels` optional.